# Applied Computational Genomics  
Winter Trimester 2024

## Course Description

The goal of this course is to initiate bench scientists on the Ph.D. track to master basic skills in bioinformatics and computational genomics. These skills are a prerequisite to a wide range of fields in modern molecular biology and are often overlooked. During this course we will acquire basic computer literacy needed to perform analyses of next generation sequencing experiments in all their modern forms. We will cover basic unix, project management with version control, principal components and clustering, visualization and color, single cell technologies, spatial omics, microbiome, and a brief survey of the fundamentals of Machine Learning. We will empower students with basic skills to advocate for and analyze their own data.

### Course Director  
Dennis Hazelett, PhD  
Assistant Professor in Computational Biomedicine  
G-577 Pacific Design Center, Green bldg.  
West Hollywood  
Office: (424) 315-4412  
Dennis.hazelett at csmc.edu
### Instructors and guest lecturers:  
Simon Coetzee (simon.coetzee at csmc.edu)  
Ivan Vujkovic-Cvijin (Ivan.Vujkovic-Cvijin at csmc.edu)   
Kyoung-Jae Won (KyoungJae.Won at cshs.org)  
Ryan Urbanowicz

Time: 9:00 AM – 10:30 AM Tuesdays (outlook calendar invitation)

Location: Picasso Room PDCG

Location of pre-recorded lectures

Some lectures in the course may be taught using the “inverted” method, in which pre-recorded lectures are augmented by structured class discussions and demonstrations. These will be made available at least 1 week prior to the relevant topic week.

Office hours: by appointment

### Prerequisites

There are no prerequisites to take this class other than the requirements of admission to Cedars-Sinai’s graduate program.

### Class Format

Classes are held every Tuesday from 9:00-10:30 AM in Picasso on the 5th floor of PDC green building (PDC-G). Classes are graded 20% attendance and contribution, 80% homework assignments and problem sets. Many lectures will be done in follow-along demonstration or mixed lecture and demonstration style. At the beginning of each lecture we will spend time addressing important questions from the 5-minute feedback (see Expectations, below). Students will be provided with a virtual machine for the duration of the course that has sufficient resources to complete any assignments. If desired each student may use their own resources with some limited support from course staff. There will be at most one homework assignment per lecture, often with required reading. Every homework assignment is due 1 week from the date of assignment at midnight. Office hours will be granted by availability and first come first serve EXCEPT ON THE DUE DATE. Early in the class, students will learn to use github, a critically important skill for maintaining open science publication standards. For the remainder of the course, completed homework assignments will be pushed to each student’s private repository (“repo”) on the github website which is to be shared with the course instructors only.

### Expectations

Attendance is required no exceptions. You may obtain permission ahead of time or with extenuating circumstances after the fact from the graduate school (email to Emma Yates-Kassler). Homework assignments are required on time (see schedule below) regardless of attendance or for half credit one lecture late. Unexcused absences result in an incomplete grade.

After each lecture, a link will be posted to provide feedback. Attendance will be recorded by the completion of the "5-minute feedback". There will be three questions, with space for brief answers:

•	What was the most important concept or take-home message you learned from today's class?  
•	What concepts or topics from today's class do you not understand well enough?  
•	What suggestions do you have for improving or enhancing today's class?

Frequently asked questions or topics of confusion will be addressed in the subsequent lecture.

### Course Evaluation

A letter grade is assigned at the end of the course according to the following breakdown:
80% assignments
20% lecture attendance and feedback

### Assignments

There is one assignment for each lecture and no exams. Late assignments will be given a maximum of 50% of full credit up to one lecture after the original due date. When homework is not assigned, points will be given for attendance. Additional readings may be indicated on the syllabus. Each question set is worth 80 points and will be graded based on completeness. Links to lecture slides and readings will be provided in the syllabi below.

### Academic Honesty:

It is expected that students in a PhD program produce only original work representing their own ideas. This extends to use of published or public sources as well as the work of other students and colleagues. Instances of likely (not proven) plagiarism of code or other intellectual property or prose will be submitted to the graduate program curriculum committee for potential sanctions (read the student handbook for further details), which could include dismissal from the program. 

### Students with Disabilities

We will provide reasonable accommodations to students with disabilities. If this applies to you, please make an appointment with the course instructors (Dennis & Simon) as soon as possible to discuss your needs.

### Class Schedule

#### 01/09/2023 (1) [Visualization and color](https://docs.google.com/presentation/d/1OsDmRKvbfmQnROLk8DDxSWyR-TR8A-hwnsXZf0Xy-JA/edit?usp=sharing)  

##### Resources:  

[The Visual Display of Quantitative Information by Edward Tufte](https://www.edwardtufte.com/tufte/books_vdqi)

#### 01/16/2023 (2) [Course intro. Unix skills.](https://docs.google.com/presentation/d/1gum_kmylz6Nw653mR31H6zasVyg3hyDlj52ZyI3Ixws/edit?usp=sharing)  

##### Resources for learning basic Linux:  
[linuxjourney.com](https://linuxjourney.com)  
[Linux for Hackers by OccupyTheWeb (Author)](https://www.amazon.com/Linux-Basics-Hackers-Networking-Scripting/dp/1593278551)  

#### 01/23/2023 (3) [Unix Skills. Version Control & Git I](https://docs.google.com/presentation/d/1ld3uKluegJVal4lo062kN8CoFmpFxZA3ird1o2Xughs/edit?usp=sharing)  

##### Resources/links for git:  
[Atlassian Git course on Coursera](https://www.coursera.org/learn/version-control-with-git)  
[git documentation](https://git-scm.com/docs)  

#### 01/30/2023 (4) [Version control & Git II](https://docs.google.com/presentation/d/1uRXryh2TLnAxYgxDNO8o5d_gbp0kIApa5HS1MkAjxb4/edit?usp=sharing)  

##### Resources:  
[Article: What is version control?](https://resources.pcb.cadence.com/blog/what-is-a-version-control-system)  
[Poetry Foundation: "The Jabberwocky" (by Lewis Carroll)](https://www.poetryfoundation.org/poems/42916/jabberwocky)  

#### 02/06/2023 (5) [Searching files with grep, cut, and sort.](https://junkdnalab.github.io/acg_2024/lectures/lecture5/file_searching_lecture5.html)  

##### Resources:  
[The best video to introduce a computational biology course.](https://twitter.com/PhillipCompeau/status/1674131281679097856)  

#### 02/13/2023 (6) [Genome Arithmetic with bedtools + advanced grep and awk](https://junkdnalab.github.io/acg_2024/lectures/lecture6/file_searching.html#/)

##### Homework
[Homework for bedtools and GenomicRanges](https://junkdnalab.github.io/acg_2024/assignments/assignment_05.md)

[feedback form](https://docs.google.com/forms/d/e/1FAIpQLSeVyqBMXe086HcmxI0v5yeMhG3P41t2Gwsze6q5e9PXZV_Gdw/viewform?usp=sf_link)

![compiling](https://imgs.xkcd.com/comics/compiling.png)

#### 02/20/2023 (7) [NGS and File Formats](https://junkdnalab.github.io/acg_2024/lectures/lecture7/file_formats.html#/)

#### 02/20/2023 (8) [Sprint Through scRNA Analysis](https://junkdnalab.github.io/acg_2024/lectures/lecture8/scrna_sprint.html)

##### Analysis File

[analysis_lecture8.R](https://junkdna.github.io/acg_2024/files/analysis_lecture8.R)

##### Homework
[Homework for File Formats](https://junkdnalab.github.io/acg_2024/assignments/assignment_06.md)

02/27/2023 (8) Single cell I  

#### 03/05/2023 (8) Single cell II  

[PCA short lecture (slides)](https://docs.google.com/presentation/d/1g0hFymDUAOooExDDpyZzSk5NA6TssYy-L0GXDEXFS1w/edit?usp=sharing).  

#### 03/12/2023 (9) Single Cell III  

[Analysis File](https://github.com/Junkdnalab/acg_2024/blob/main/files/analysis_lecture10.R)

[split_seurat.rda](https://cloud.coetzee.me/s/sqqEtpZP7wmJkb6)

[integration_speedup.rda](https://cloud.coetzee.me/s/kNJQaQ9EQNYAawq)

#### 03/19/2023 (10) Spatial Omics I  

[Spatial Transcriptomics (slides)](https://docs.google.com/presentation/d/12JKPp_P1IWG0nzy9O0aFbKNnfm2adO5-/edit#slide=id.p1).  
[Visium dataset (google drive link)](https://drive.google.com/file/d/1mg-PnQr-6yprEd18kCbmNa4OO7q9Zl2B/view). 

#### 03/26/2023 (11) Spatial Omics II  
[Spatial Omics II - Course Pack](https://github.com/Junkdnalab/acg_2024/blob/main/files/spatial_omics.zip)

04/02/2023 (11) Introduction to microbiome analysis  
04/09/2023 (12) Microbiome analysis I  
04/16/2023 (13) Microbiome analysis II  
